hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/ChunkInputStream.java

```diff
@@ -0,0 +1,536 @@
+   * @throws IOException if there is an I/O error while performing the call
+   * to Datanode
+   */
+  private synchronized void readChunkFromContainer(int len) throws IOException {
+
+    List<DatanodeDetails> excludeDns = null;
+    ByteString byteString;
+    List<DatanodeDetails> dnList = getDatanodeList();
+
+    // index of first byte to be read from the chunk
+    long startByteIndex;
+    if (chunkPosition >= 0) {
+      // If seek operation was called to advance the buffer postion, the
+      // chunk should be read from that postion onwards.
+      startByteIndex = chunkPosition;
+    } else {
+      // Start reading the chunk from the last chunkPosition onwards.
+      startByteIndex = bufferOffset + bufferLength;
+    }
+
+    if (verifyChecksum) {
+      // Update the bufferOffset and bufferLength as per the checksum
+      // boundary requirement.
+      computeChecksumBoundaries(startByteIndex, len);
+    } else {
+      // Read from the startByteIndex
+      bufferOffset = startByteIndex;
+      bufferLength = len;
+    }
+
+    // Adjust the chunkInfo so that only the required bytes are read from
+    // the chunk.
+    final ChunkInfo adjustedChunkInfo = ChunkInfo.newBuilder(chunkInfo)
+        .setOffset(bufferOffset)
+        .setLen(bufferLength)
+        .build();
+
+    while (true) {
+      List<DatanodeDetails> dnListFromReadChunkCall = new ArrayList<>();
+      byteString = readChunk(adjustedChunkInfo, excludeDns,
+          dnListFromReadChunkCall);
+      try {
+        if (byteString.size() != adjustedChunkInfo.getLen()) {
+          // Bytes read from chunk should be equal to chunk size.
+          throw new IOException(String
+              .format("Inconsistent read for chunk=%s len=%d bytesRead=%d",
+                  adjustedChunkInfo.getChunkName(), adjustedChunkInfo.getLen(),
+                  byteString.size()));
+        }
+
+        if (verifyChecksum) {
+          ChecksumData checksumData = ChecksumData.getFromProtoBuf(
+              chunkInfo.getChecksumData());
+          int checkumStartIndex =
+              (int) (bufferOffset / checksumData.getBytesPerChecksum());
+          Checksum.verifyChecksum(byteString, checksumData, checkumStartIndex);
+        }
+        break;
+      } catch (IOException ioe) {
+        // we will end up in this situation only if the checksum mismatch
+        // happens or the length of the chunk mismatches.
+        // In this case, read should be retried on a different replica.
+        // TODO: Inform SCM of a possible corrupt container replica here
+        if (excludeDns == null) {
+          excludeDns = new ArrayList<>();
+        }
+        excludeDns.addAll(dnListFromReadChunkCall);
+        if (excludeDns.size() == dnList.size()) {
+          throw ioe;
+        }
+      }
+    }
+
+    buffers = byteString.asReadOnlyByteBufferList();
+    bufferIndex = 0;
+    allocated = true;
+
+    // If the stream was seeked to postion before, then the buffer
```

Comment:
Minor NIT: postion -> position.
Need to be changed in multiple places.

Reply:
done